File: commands/src/main/java/kaap/veiko/debuggerforker/commands/commandsets/event/events/ClassUnloadEvent.java
package kaap.veiko.debuggerforker.commands.commandsets.event.events;

import com.google.auto.value.AutoValue;

import kaap.veiko.debuggerforker.types.DataReadException;
import kaap.veiko.debuggerforker.types.DataReader;
import kaap.veiko.debuggerforker.types.DataWriter;
import kaap.veiko.debuggerforker.types.jdwp.EventKind;

@AutoValue
public abstract class ClassUnloadEvent extends VirtualMachineEvent {
  public abstract int getRequestId();
  public abstract String getSignature();

  public EventKind getEventKind() {
    return EventKind.CLASS_UNLOAD;
  }

  public static ClassUnloadEvent create(int requestId, String signature) {
    return new AutoValue_ClassUnloadEvent(requestId, signature);
  }

  public static ClassUnloadEvent read(DataReader reader) throws DataReadException {
    return create(
        reader.readInt(),
        reader.readString()
    );
  }

  @Override
  public void write(DataWriter writer) {
    writer.writeType(getEventKind());
    writer.writeInt(getRequestId());
    writer.writeString(getSignature());
  }
}

File: commands/src/main/kotlin/kaap/veiko/debuggerforker/commands/commandsets/event/events/ClassUnloadEvent.kt
package kaap.veiko.debuggerforker.commands.commandsets.event.events

import kaap.veiko.debuggerforker.types.DataReader
import kaap.veiko.debuggerforker.types.DataWriter
import kaap.veiko.debuggerforker.types.jdwp.EventKind

data class ClassUnloadEvent(
        val requestId: Int,
        val signature: String
) : VirtualMachineEvent() {

    override fun write(writer: DataWriter?) = writer!!.run {
        writeType(EventKind.CLASS_UNLOAD)
        writeInt(requestId)
        writeString(signature)
    }

    companion object {
        @JvmStatic
        fun read(reader: DataReader) = ClassUnloadEvent(
                reader.readInt(),
                reader.readString()
        )
    }
}